File: 3307-find-the-maximum-sum-of-node-values/3307-find-the-maximum-sum-of-node-values.cpp
class Solution {
public:
    long long maximumValueSum(vector<int>& nums, int k, vector<vector<int>>& edges) {
        int cnt = 0;
        long long totalSum = 0;
        int positive_Min = INT_MAX;
        int negative_Max = INT_MIN;

        for(int nodeValue : nums) {
            int nodeAfterOp = nodeValue ^ k;
            totalSum += nodeValue;
            int netChange = nodeAfterOp - nodeValue;

            if(netChange > 0) {
                positive_Min = min(netChange, positive_Min);
                totalSum += netChange;
                cnt += 1;
            } else {
                negative_Max = max(negative_Max, netChange);
            }
        }

        if(cnt % 2 == 0) {
            return totalSum;
        }

        return max(totalSum - positive_Min, totalSum + negative_Max);
    }
};

File: 3307-find-the-maximum-sum-of-node-values/3307-find-the-maximum-sum-of-node-values.go
func maximumValueSum(nums []int, k int, edges [][]int) int64 {
    sum := int64(0)
    positiveMin := 1000000000
    negativeMax := -1000000000
    cnt := 0
    
    for _, num := range nums {
        flipped := num ^ k 
        profit := flipped - num
        sum += int64(num)
        if profit > 0 {
            sum += int64(profit)
            positiveMin = min(positiveMin, profit)
            cnt++
        } else {
            negativeMax = max(negativeMax, profit)
        }
    } 

    if cnt % 2 == 0 {
        return sum
    }
    return max(sum + int64(negativeMax), sum - int64(positiveMin))
}